### PR TITLE
Extract files from encrypted pcap files

### PR DESCRIPTION
**Pull request by jochman**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/26265

## Description
Added two arguments:
- wpa_password: for password-protected wpa-pwd files.
- rsa_decrypt_key_entry_id: for rsa encrypted files.

Needed to add Magic to the pcap-miner docker (That's why it's new)

## Minimum version of Demisto
- [ ] 4.5.0
- [ ] 5.0.0
- [ ] 5.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 1]
```diff
@@ -1,64 +1,155 @@
...
+        stderr: STDERR pipe
 
+    Raises:
+        DemistoException if returncode is different than 0
     """
-    file_entry_id = args.get('entry_id', '')
+    process = subprocess.Popen(args, stdout=stdout, stderr=stderr)
+    stdout_data, stderr_data = process.communicate()
+    if process.returncode != 0:
+        raise DemistoException(f'Error returned from tshark command: {process.returncode=}\n {stderr_data=}')
```
Comment (yaakovi) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:39]:
```suggestion
        raise DemistoException(f'Error returned from tshark command: {process.returncode}\n {stderr_data!r}')
```

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 2]
```diff
@@ -1,64 +1,155 @@
...
+
+    # For debugging purposes
+    return stdout_data, stderr_data
 
-    res = demisto.executeCommand('getFilePath', {'id': file_entry_id})
-    if is_error(res):
-        raise Exception(f'Failed to get the file path for entry: {file_entry_id} - {get_error(res)}')
 
-    return res[0]['Contents']['path']
+def filter_files(root: str, files: List[str], /,
```
Comment (yaakovi) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:45]:
```suggestion
def filter_files(root: str, files: List[str],
```

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 3]
```diff
@@ -1,64 +1,155 @@
...
+            if extensions_inclusive_or_exclusive == InclusiveExclusive.INCLUSIVE and f_ext not in extensions:
+                files.remove(file)
+            # Exclude extensions, don't take those files.
+            elif extensions_inclusive_or_exclusive == InclusiveExclusive.EXCLUSIVE and f_ext in extensions:
+                files.remove(file)
+    return files
+
+
+def upload_files(
+        file_path: str, dir_path: str, /,
```
Comment (yaakovi) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:89]:
```suggestion
        file_path: str, dir_path: str,
```

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 4]
```diff
@@ -82,23 +173,38 @@ def upload_files(dir_path: str, file_path: str) -> Union[CommandResults, str]:
...
+        types: Optional[str] = None,
+        types_inclusive_or_exclusive: Optional[str] = 'inclusive',
+        extensions: Optional[str] = None,
+        extensions_inclusive_or_exclusive: str = 'inclusive',
+        limit: Optional[str] = '5',
+):
+    with tempfile.TemporaryDirectory() as dir_path:
+        try:
+            file_path, file_name = get_file_path_from_id(entry_id)
+            cert_path, _ = get_file_path_from_id(rsa_decrypt_key_entry_id)
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:194]: i think it makes sense to have the condition here to check if `rsa_decrypt_key_entry_id` was provided then get its file path, rather than returning `None, None` from the func

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 5]
```diff
@@ -1,64 +1,155 @@
...
-    """
+class InclusiveExclusive:
+    INCLUSIVE: str = 'inclusive'
+    EXCLUSIVE: str = 'exclusive'
 
-    Args:
-        args: argument from demisto
 
-    Returns: path of pcap file
+def get_file_path_from_id(entry_id: Optional[str] = None):
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:17]: with regards to my comments about that func in `main`, please add return type hint

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 6]
```diff
@@ -1,64 +1,155 @@
...
         Extracted files to download
 
     """
-
-    process = subprocess.Popen(['tshark', '-r', f'{file_path}', '--export-objects', f'http,{dir_path}',
-                                '--export-objects', f'smb,{dir_path}', '--export-objects', f'imf,{dir_path}',
-                                '--export-objects', f'tftp,{dir_path}', '--export-objects', f'dicom,{dir_path}'],
-                               stdout=subprocess.PIPE, stderr=subprocess.PIPE)
-    process.communicate()
+    if kwargs is not None:
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:115]: why is this needed?

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 7]
```diff
@@ -1,64 +1,155 @@
...
+    Raises:
+        DemistoException if returncode is different than 0
     """
-    file_entry_id = args.get('entry_id', '')
+    process = subprocess.Popen(args, stdout=stdout, stderr=stderr)
+    stdout_data, stderr_data = process.communicate()
+    if process.returncode != 0:
+        raise DemistoException(f'Error returned from tshark command: {process.returncode}\n {stderr_data!r}')
+
+    # For debugging purposes
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:41]: can you elaborate why is this needed?

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 8]
```diff
@@ -1,64 +1,155 @@
...
-    res = demisto.executeCommand('getFilePath', {'id': file_entry_id})
-    if is_error(res):
-        raise Exception(f'Failed to get the file path for entry: {file_entry_id} - {get_error(res)}')
 
-    return res[0]['Contents']['path']
+def filter_files(root: str, files: List[str],
+                 types: Optional[Set[str]] = None,
+                 extensions: Optional[Set[str]] = None,
+                 types_inclusive_or_exclusive: Optional[str] = None,
+                 extensions_inclusive_or_exclusive: Optional[str] = None):
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:49]: add return type hinnt

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 9]
```diff
@@ -1,64 +1,155 @@
...
+            mime_type = magic_mime.from_file(os.path.join(root, file))
+            # Inclusive types, take only the types in the list.
+            if types_inclusive_or_exclusive == InclusiveExclusive.INCLUSIVE and mime_type not in types:
+                files.remove(file)
+            # Exclusive types, don't take those files.
+            elif types_inclusive_or_exclusive == InclusiveExclusive.EXCLUSIVE and mime_type in types:
+                files.remove(file)
+        if extensions:
+            # strip `.` from extension
+            extensions = set([extension.split(".")[-1] for extension in extensions])
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:76]:
```suggestion
            extensions = set([extension.split('.')[-1] for extension in extensions])
```

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 10]
```diff
@@ -82,23 +173,38 @@ def upload_files(dir_path: str, file_path: str) -> Union[CommandResults, str]:
...
-        return_results(upload_files(dir_path, file_path))
-
-    except Exception as e:
-        return_error(f'Failed to execute PcapFileExtracor. Error: {str(e)}')
-    finally:
-        if dir_path:
-            shutil.rmtree(dir_path)
+        return results
+    else:
+        raise DemistoException('No files found in path.')
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:178]: this should be raised as an exception, but a regular entry

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 11]
```diff
@@ -1,64 +1,155 @@
...
+    run_command(command)
 
     context = []
-
     md5 = hashlib.md5()
     sha1 = hashlib.sha1()
     sha256 = hashlib.sha256()
     for root, _, files in os.walk(dir_path):
-        if len(files) == 0:
+        # Limit the files list to minimum
```
Comment (Itay4) [inline Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:137]: to minimum? whats this comment for?

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 12]
```diff
@@ -1,64 +1,155 @@
...
-import shutil
-import subprocess
-import hashlib
+
+import demistomock as demisto
 
 
-def get_pcap_path(args: Dict) -> str:
-    """
+class InclusiveExclusive:
```
[the remark this answers is not in the document]
Comment (jochman) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:12]: It's kind of enum. we should not use globals

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py [hunk 13]
```diff
@@ -82,23 +173,38 @@ def upload_files(dir_path: str, file_path: str) -> Union[CommandResults, str]:
...
+
+def main(
+        entry_id: str,
+        wpa_password: Optional[str] = None,
+        rsa_decrypt_key_entry_id: Optional[str] = None,
+        types: Optional[str] = None,
+        types_inclusive_or_exclusive: Optional[str] = 'inclusive',
+        extensions: Optional[str] = None,
+        extensions_inclusive_or_exclusive: str = 'inclusive',
+        limit: Optional[str] = '5',
```
[the remark this answers is not in the document]
Comment (jochman) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:189]: it can get None if it's not in the args. So should be typed as Optional.

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (Itay4) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:12]: const globals are fine..

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Itay4) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:189]: theres a default value of `5`, it will never be `None`

File: Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jochman) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:12]: so it's an enum :)

[the remark this answers is not in the document]
Comment (Itay4) [reply Packs/PcapAnalysis/Scripts/PcapFileExtractor/PcapFileExtractor.py:12]: i still think its an overkill.. i mean the class name is literally the only 2 options there are.. looks weird
but if you insist, at least make it a proper enum https://docs.python.org/3/library/enum.html

**Comment by kirbles19** [comment]
@jochman Done.
